BUG GH49716 Improve type hint accuracy for pandas.read_sql_query

xuganyu96 opened the pull request:
- [ ] closes #49716
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

xuganyu96 committed:
removed inaccurate type hints

phofl commented on pandas/io/sql.py:
```diff
@@ -307,7 +307,7 @@ def read_sql_query(
     con,
     index_col: str | list[str] | None = ...,
     coerce_float=...,
-    params: list[str] | dict[str, str] | None = ...,
```
Please type this as needed instead of removing the types altogether

xuganyu96 replied:
Thank you for the feedback! Based on the doc string under `read_sql_query`, `params` can be "list, tuple or dict, optional." Would it make more sense to revise the type hints to `list | tuple | dict | None`?

phofl replied:
ist list[str] incorrect? Otherwise, there is no reason to get less specific. Even if str is not correct, we would want to narrow it down, because something like list[DataFrame] is certainly not correct

xuganyu96 replied:
`list[str]` is not so much "incorrect" as it is incomplete. I am not sure if there is a finite set of types that can be passed into `params` as values because it seems like each database client library can implement its own thing. Maybe there is something I can find on Python DB API v2 (PEP249) to narrow down the set of acceptable data types. Will return to this PR with more information when I find it.

xuganyu96 replied:
After some investigation, I stand by my original proposal that `params`'s type hint should be generically `list | tuple | dict | None` without adding additional constraint on what kind of values can go into the list, tuple, or dictionary.

In fact, SQLAlchemy defined type hints for `params` using a `Any` in many places, and I find it unreasonable to restrict the type hints in `pandas` further than what SQLAlchemy did ([reference](https://github.com/sqlalchemy/sqlalchemy/blob/3fc6c40ea77c971d3067dab0fdf57a5b5313b69b/lib/sqlalchemy/engine/interfaces.py#L224)):

```python
# From SQLALchemy's source code
_CoreSingleExecuteParams = Mapping[str, Any]
_MutableCoreSingleExecuteParams = MutableMapping[str, Any]
_CoreMultiExecuteParams = Sequence[_CoreSingleExecuteParams]
_CoreAnyExecuteParams = Union[
    _CoreMultiExecuteParams, _CoreSingleExecuteParams
]

_DBAPISingleExecuteParams = Union[Sequence[Any], _CoreSingleExecuteParams]

_DBAPIMultiExecuteParams = Union[
    Sequence[Sequence[Any]], _CoreMultiExecuteParams
]
_DBAPIAnyExecuteParams = Union[
    _DBAPIMultiExecuteParams, _DBAPISingleExecuteParams
]
_DBAPICursorDescription = Tuple[
    str,
    "DBAPIType",
    Optional[int],
    Optional[int],
    Optional[int],
    Optional[int],
    Optional[bool],
]

_AnySingleExecuteParams = _DBAPISingleExecuteParams
_AnyMultiExecuteParams = _DBAPIMultiExecuteParams
_AnyExecuteParams = _DBAPIAnyExecuteParams
```

Also, `list[DataFrame]` could actually be a valid input for `params`. Here is a depraved example using `sqlite3.register_adapters`:

```python
import sqlite3
import pandas as pd

df = pd.DataFrame({
    "nums": [1, 2, 3],
    "english": ["one", "two", "three"],
})

def adapt_dataframe(df: pd.DataFrame):
    return f"<DataFrame shape={df.shape}>"

sqlite3.register_adapter(pd.DataFrame, adapt_dataframe)

with sqlite3.connect(":memory:") as conn:
    result = pd.read_sql_query(
        sql="SELECT ?, ?, ?;",
        con=conn,
        params=[df, df, df],
    )
    print(result)

    result = pd.read_sql_query(
        sql="SELECT :v, :u, :w;",
        con=conn,
        params={
            "v": df,
            "u": df,
            "w": df,
        }
    )
    print(result)
```

WillAyd replied:
I agree with @phofl that we should stay as strict as possible, understanding that it may be a looser type check upstream. If there is a legitimate use case to parametrize with something other than str I think can do that down the road. Its OK to be more strict than the upstream type

xuganyu96 replied:
@WillAyd @phofl 
In the comment above it should have been made clear already that it is possible to use `register_adapter` (which is present in `sqlite3` and `psycopg2`, if not more client libraries) to parametrize any kind of type (the example demonstrated feeding a list or a dictionary of `pd.DataFrame` into `params`)

While I can see that this is probably not a common use case, `str` is definitely not the only use case.

xuganyu96 committed:
Updated type hints of params to be consistent with docstring

xuganyu96 committed:
Fixed default argument styling

xuganyu96 commented:
Would it be necessary for me to add to the change list for this PR?

xuganyu96 committed:
Merge branch 'main' of github.com:xuganyu96/pandas into read-sql-query-params-type-hints

xuganyu96 committed:
Merge branch 'main' into read-sql-query-params-type-hints

xuganyu96 committed:
Added entry to whatsnew/v1.5.3.rst

xuganyu96 committed:
restricts key types for when dictionary is passed to params

xuganyu96 committed:
Fixed doc style

xuganyu96 commented:
@phofl @WillAyd I have added description of what this PR tries to fix to `doc/source/whatsnew/v1.5.3.rst`, and I have modified the type hints to strict the keys to be strings for when a dictionary is passed into `params`.

I am not sure what test cases to write for a type hint change. It would be great if you can clarify. Thank you.

xuganyu96 committed:
Removed invalid directive
